mongo/integration/server_selection_prose_test.go

```diff
@@ -19,11 +20,107 @@ import (
 	"go.mongodb.org/mongo-driver/mongo/description"
 	"go.mongodb.org/mongo-driver/mongo/integration/mtest"
 	"go.mongodb.org/mongo-driver/mongo/options"
+	"golang.org/x/sync/errgroup"
 )
 
+type saturatedConnections map[uint64]bool
+
+// saturatedHosts is used to maintain information about events with specific host+pool combinations.
+type saturatedHosts map[string]saturatedConnections
+
+func (set saturatedHosts) has(host string, connectionID uint64) bool {
+	var b bool
+	if h := set[host]; h != nil {
+		b = h[connectionID]
+	}
+	return b
+}
+
+func (set saturatedHosts) add(host string, connectionID uint64) {
+	if set[host] == nil {
+		set[host] = make(saturatedConnections)
+	}
+	set[host][connectionID] = true
+}
+
+// isSaturated returns true when each client on the cluster URI has a tolerable number of ready connections.
+func (set saturatedHosts) isSaturated(mt *mtest.T, tolerance uint64) bool {
+	for _, host := range options.Client().ApplyURI(mtest.ClusterURI()).Hosts {
+		if cxns := set[host]; cxns == nil || uint64(len(cxns)) < tolerance {
+			return false
+		}
+	}
+	return true
+}
+
+// awaitSaturation uses CMAP events to ensure that the client's connection pools for N-mongoses have been saturated.
+// The qualification for a host to be "saturated" is for each host on the client to have a tolerable number of ready
+// connections.
+func awaitSaturation(ctx context.Context, mt *mtest.T, monitor *monitor.TestPoolMonitor, tolerance uint64) {
+	errs, ctx := errgroup.WithContext(ctx)
+	done := make(chan struct{}, 1)
+	errs.Go(func() error {
+		set := make(saturatedHosts)
+		for !set.isSaturated(mt, tolerance) {
+			if err := mt.Coll.FindOne(ctx, bson.D{}).Err(); err != nil {
+				return err
+			}
+			monitor.Events(func(evt *event.PoolEvent) bool {
+				// Add host only when the connection is ready for use.
+				if evt.Type == event.ConnectionReady && !set.has(evt.Address, evt.ConnectionID) {
+					set.add(evt.Address, evt.ConnectionID)
+				}
```

Comment:
Optional: Is the `add` function idempotent? If so, consider removing the `has` function because the behavior should be identical with and without the `set.has` check above.

Reply:
Wow great point!